### ISSUE TITLE
SQL database backup/restore does not include SQL procedures/functions

### ISSUE THREAD

**Issue by jnbarchan** [hezmondo/Jinn#124]
We recently started writing SQL `FUNCTION`s as part of Jinn (`utilities/scripts/databasecode.sql`).  If restoring a database from a working system to a brand new one, these do not get restored.

MySQL has a default that `PROCEDURE`/`FUNCTION`s are *not* included in backup.  I forgot to update our `utilities/scripts/backup-database` to pass an argument for this.

Changing backup to pass `--routines` for this.

Restore should then work properly.  Note that this now requires the (SQL) user restoring to have:

> This option requires the `SELECT` privilege for the `mysql.proc` table.

**Comment by jnbarchan** [comment]
Major headache: as I half-suspected, with that option `mysqldump` produces lines like:

`CREATE DEFINER=`jon`@`%` FUNCTION`

MySQL has this *ridiculous* behaviour of *insisting* on putting the "definer" name in, which makes backups unsuitable for restoring elsewhere, and they don't seem to care that this makes it unusable....

**Comment by LaurenceLeo** [comment]
Is it possible to make Jinn run a SQL query to check if the procedures exist on startup? Then if they don't run a bat/bash file like your `databaseupgrade ` scripts to add them? 

This way we wouldn't have to worry about `mysqldump` being an ass.

**Comment by jnbarchan** [comment]
We wish to continue to use `mysqldump`, not write out own.  In that light, I have decided *not* to change the backup to include procedures/functions, at least for now.

The problem of not having the functions in the restored backup only occurs if you are restoring to a completely new database which does not already have the current functions.  This is an infrequent situation, perhaps only arising every couple of years if, say, the whole OS or machine is changed.

Consequently we have decided that this is low priority, and a manual workaround is acceptable.

When this situation arises, rather than doing what was said above about changing the internal `database_version` to try to force an upgrade (every upgrade automatically reads in the current SQL function definitions, which are supplied in file `utilities/scripts/databasecode.sql`, because they might have been altered), the correct workaround is simply to go something like (see `utilities/scripts/upgrade-database`):
```
mysql "--host=localhost" "--user=jon" "--password=..." "--database=yoda" --skip-column-names --unbuffered -e "source utilities/scripts/databasecode.sql;"
```
